### ISSUE TITLE
Worker pool - recursively adding items?

### ISSUE THREAD

**Issue by artjomsimon** [project-iris/iris#54]
Hi everyone,

I've played a bit with the worker pool, which works fine when I'm adding tasks to it which it can crunch on.
However, while experimenting with a piece of C code using OpenMP and trying to see how it would look in Go, I understand that the way the pool is designed using a FIFO queue, there's a case where the pool deadlocks: Creating tasks recursively, while waiting for a result.
What seems like an academic, contrieved example, actually is  (an irregular algorithm where n tasks are created recursively for benchmarking purposes)

Here's an example: http://play.golang.org/p/eSeh1ijy1g (I've simply copy&pasted the work pool package into the play sandbox, with an example at the top, for the sake of demonstration)

Basically, the resursively called func is scheduled, blocking a place in the queue, waiting for a result that never comes, because the last recursion step that actually would unwind the chain of recursion, never gets scheduled, because the pool's queue is already full.

You can increase the pool size, of course, healing symptoms, only to run into the problem when the recursion is deeper than the size of the pool.

I understand that even if this kind of problem isn't that common, I thought this is a good place to ask, because I'm sure someone has already thought how this could be solved elegantly in Go and can point me in the right direction.

Thank you,
Artjom
